### DISCUSSION TITLE
Connecting line (i.e. text drawn 'over' the stroke of a circle)

### DISCUSSION THREAD

**Discussion by flo80** [ookami-kb/flutter_arc_text#14] [Ideas]
For an app we are working on, we wanted to have a visible circle with arced text (middle alignment). Since we paint this over a custom background we cannot just paint text with black background but needed to draw the remaining arc of the circle. 

If this is an interesting feature for integration please let me know and I can create a PR. 
You can have a look in my [fork](https://github.com/flo80/flutter_arc_text/tree/feature/connecting-line) (note: still based on last non null-safe version of arc_text, will probably upgrade it in the coming days)

**Comment by ookami-kb** [comment]
I'm not sure that I understand what you mean. Can you please provide a screenshot?

**Reply by flo80** [reply]
Sure. 

This is how it looks when you draw text and have box decoration circle 
![image](https://user-images.githubusercontent.com/12986622/113979998-46546200-9846-11eb-90a5-581d2c6a23ca.png)

With the 'connecting line' (not best naming) you can do things like these
![image](https://user-images.githubusercontent.com/12986622/113980168-71d74c80-9846-11eb-99c8-11e13d0fb0c8.png)
![image](https://user-images.githubusercontent.com/12986622/113980220-80256880-9846-11eb-98e5-72374a7afb3b.png)

It might be a slightly niche feature but if you need it I did not find another easy way to do it

**Reply by monsieurtanuki** [reply]
Hi there!
In order to do something similar, I used the computed angle of an arc text.
Then I was able to draw anything (e.g. a circle arc, or 3 icons) from the end of the arc text to its beginning. I don't think it should be limited to dash circle arc, therefore the `connectingLine*` parameters don't look like a good idea to me.
I don't really know how this library evolved since the day I had to fork, but the root of flexibility is that: to be able to know the angle of an arc text. After or before (e.g. for a background) painting it.

**Reply by flo80** [reply]
How did you implement this, by forking and changing the paint method or is there a way from the ‘outside’ / parent widget to get the angle information?

**Reply by ookami-kb** [reply]
I agree with @monsieurtanuki that adding connecting line seems too limited for a library.

Right now you can use `ArcTextPainter` to customize the behavior. It's not super-convenient at the moment (so we probably can think about improving customization support), but at least it should be easier than maintaining the fork.

You can use something like this:

```dart
class _CustomPainter extends CustomPainter {
  final _startAngle = -pi / 2;

  final _paint = Paint()
    ..style = PaintingStyle.stroke
    ..strokeWidth = 2
    ..color = Colors.red;

  late final ArcTextPainter _painter = ArcTextPainter(
    radius: 100,
    text: ' Custom painter text with ArcTextPainter ',
    textStyle: const TextStyle(color: Colors.black, fontSize: 16),
    placement: Placement.middle,
    initialAngle: _startAngle,
  );

  @override
  void paint(Canvas canvas, Size size) {
    final center = Offset(size.width / 2, size.height / 2);
    final path = Path()
      ..arcTo(
        Rect.fromCenter(center: center, width: size.width, height: size.height),
        _painter.getFinalAngle() - pi / 2 + _startAngle,
        2 * pi - _painter.getFinalAngle(),
        true,
      );
    canvas.drawPath(path, _paint);
    _painter.paint(canvas, center);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}

final widget = SizedBox(
  width: 200,
  height: 200,
  child: CustomPaint(
    painter: _CustomPainter(),
  ),
);
```

<img width="352" alt="image" src="https://user-images.githubusercontent.com/853356/114154497-c222dd00-9920-11eb-968b-57fd31580e9b.png">

**Reply by flo80** [reply]
Hm, really not the most convenient way. What about adding a callback function which gets called in the painter and provides canvas, size, angle/ length info?

**Reply by ookami-kb** [reply]
Yeah, something like visitor pattern for the painter makes sense.

**Reply by monsieurtanuki** [reply]
I think a callback function wouldn't be a good option because:
1. it would prevent the use of the angle _before_ the text is drawn - e.g. for an arc text specific background
2. IMO the notion of "callback" is attached to some asynchronous code to be triggered when some other asynchronous code ends but we have no idea when - in our case we know exactly when (at the end of the arc text paint), and we don't want the second code to be played asynchronously

I don't know about the visitor pattern.

The goals are:
1. I want to display an arc text (we can do that today, perfect!)
2. I want to know the start and end angle of an arc text without having to paint it
 
A solution could be just a method with tons of code reuse: "normal" users that just want to display an arc text like today wouldn't care about it , and "advanced" users would call it before or after the arc text paints.

**Reply by monsieurtanuki** [reply]
@flo80 To answer your previous question (and as in @ookami-kb's sample code), I just use `getFinalAngle`.
Perhaps additional methods `getStartAngle(double finalAngle)` and `getEndAngle(double finalAngle)` would make things easier and we wouldn't have to use `pi/2` and stuff like that. It is possible in dart to make those methods belong to `enum StartAngleAlignment`?

**Reply by flo80** [reply]
@ookami-kb to have a simpler discussion what could work, I create PR #16 with a very simple method that allows painting before the text is drawn

@monsieurtanuki I think 'callback' was just a wrong wording, I just want a method which allows custom painting before the arc text is painted which has access to the angle information (without all the boilerplate required as in the example above)

**Reply by ookami-kb** [reply]
I was thinking about something like this: https://github.com/ookami-kb/flutter_arc_text/pull/17

This PR introduces a couple of breaking changes, but they should be simple.

**Comment by ookami-kb** [comment]
@flo80 @monsieurtanuki 

Version `0.5.0` published with the latest changes. Thanks a lot for your help!
